feat(transformer): support nullish-coalescing-operator plugin

Dunqing opened the pull request:
The implementation copy from the original implementation which removed in https://github.com/oxc-project/oxc/pull/2865.

Dunqing commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4884?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#4884** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4884?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* **#4876** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4876?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Dunqing and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

overlookmotel commented on crates/oxc_transformer/src/es2020/nullish_coalescing_operator.rs:
```diff
@@ -0,0 +1,125 @@
+
+        // skip creating extra reference when `left` is static
+        if ctx.symbols().is_static(&logical_expr.left) {
+            reference = ctx.ast.copy(&logical_expr.left);
+            assignment = ctx.ast.copy(&logical_expr.left);
```
Please god no!

If you insert this after `if logical_expr.operator != LogicalOperator::Coalesce`, then you have an owned `LogicalExpression` and can remove all the `ast.copy` calls:

```rs
// Take ownership of the `LogicalExpression`
let logical_expr = match ctx.ast.move_expression(expr) {
    Expression::LogicalExpression(logical_expr) => logical_expr.unbox(),
    _ => unreachable!(),
};
```

This is pretty horrible, but should not be so bad performance-wise. The compiler should be able to deduce that it's a `LogicalExpression` (because we already checked that above) and remove the `unreachable!()` branch (because it is indeed provably unreachable).

As it stands, this code is definite UB, as you are using `logical_expr.left` and `reference` twice. Once you take out the `ast.copy` calls, the compiler draws your attention to this.

Dunqing replied:
Yes! You're right. I missed these `ctx.copy` usages. Now I applied your suggestion. That looks good!

Dunqing commented on crates/oxc_transformer/src/es2020/nullish_coalescing_operator.rs:
```diff
@@ -0,0 +1,125 @@
+    ) -> IdentifierReference<'a> {
+        // Add `var name` to scope
+        let name = match expr {
+            Expression::Identifier(ref ident) => ident.name.as_str(),
```
I needed it before and I forgot to delete it

Dunqing commented:
@overlookmotel Thank you for reviewing this! All of the requests look good, I applied some suggestions, and the remaining I would like to do after this PR is merged.

Dunqing committed:
feat(transformer): support nullish-coalescing-operator plugin (#4884)

The implementation copy from the original implementation which removed in https://github.com/oxc-project/oxc/pull/2865.
